### PR TITLE
Load Modded Dimensions on Dedicated Server Start 

### LINKED ISSUE #8552: [1.18.2-40.0.36] Json dimensions in mods do not apply on first serverside world creation
When using a mod that has json dimensions inside, the initial creation of the world folder on a server will not contain that mod's dimension. This forces server owners to restart their server after the first time in order for the modded dimension to show up in game. Example of this can be seen using The Bumblezone mod but it also impacts many other dimension mods as well.

Bumblezone mod: https://www.curseforge.com/minecraft/mc-mods/the-bumblezone-forge

Add the mod along with its dependencies to the mods folder for a server. Make sure the world folder is deleted or non-existent on the server. Start the server. Give yourself op and then do `/execute in` and see how the autocomplete does not show the modded dimension at all. Restart the server and try again without deleting the world folder. Now the command will show the modded dimension.

After initial world creation on server:
image

After a server restart:
image

World folders:
[worldwithoutdimension.zip](https://github.com/MinecraftForge/MinecraftForge/files/8404753/worldwithoutdimension.zip)
[worldafterrestart.zip](https://github.com/MinecraftForge/MinecraftForge/files/8404765/worldafterrestart.zip)

Note how the in the level.dat file, the Data/WorldGenSettings/dimensions package is missing the modded dimension on first creation but gets it on restart.

### PR DESCRIPTION
**Pull request by Wyn-Price**
Closes #8552

I've tested with the `server_world_creation_test` mod, which has a custom dimension, dimension type, and noise settings (noise settings is to confirm that the other registries load datapacks)

<s>As fair as I can tell, `Dimension` and `DimensionType` use only the vanilla entries when starting a world on a dedicated server. This makes sense from a vanilla standpoint, as the datapack folder is inside the world folder, so if there is no world, there is no datapacks to load. However this excludes all modded dimension types/dimensions which are loaded in "datapacks" outside the world/datapacks folder
 - DimensionTypes can be registered immediately
 - Dimensions need to be registered later, as they are registered to a local variable registry before the main registry. 

If the loading of a modded dimensions/dimension type fails, I've logged a warning to the console.

Note that this code should only run on dedicated servers. The checks for `registryAccess.get() == null` or `registryResouceAccess.get() == null` do this, as they're only ever set on dedicated servers.</s>

### CODE REVIEW

File: src/test/resources/META-INF/mods.toml [hunk 1]
```diff
@@ -214,5 +214,6 @@ license="LGPL v2.1"
     modId="hide_neighbor_face_test"
 [[mods]]
     modId="living_get_projectile_event_test"
-
```
Comment (sciwhiz12) [inline src/test/resources/META-INF/mods.toml:217]: Please keep this blank line above the comment.

File: src/test/java/net/minecraftforge/debug/world/ServerWorldCreationTest.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import net.minecraftforge.event.world.BiomeLoadingEvent;
+import net.minecraftforge.fml.common.Mod;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+@Mod(ServerWorldCreationTest.MODID)
+public class ServerWorldCreationTest {
+    static final String MODID = "server_world_creation_test";
+    private static final Logger LOGGER = LogManager.getLogger(MODID);
+    public ServerWorldCreationTest(){
```
Comment (sciwhiz12) [inline src/test/java/net/minecraftforge/debug/world/ServerWorldCreationTest.java]: Formatting: braces on newlines.
```suggestion
    public ServerWorldCreationTest(
    {
```

File: src/test/java/net/minecraftforge/debug/world/ServerWorldCreationTest.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+import net.minecraft.resources.ResourceLocation;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.world.BiomeLoadingEvent;
+import net.minecraftforge.fml.common.Mod;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+@Mod(ServerWorldCreationTest.MODID)
+public class ServerWorldCreationTest {
```
Comment (sciwhiz12) [inline src/test/java/net/minecraftforge/debug/world/ServerWorldCreationTest.java]: Please provide a javadoc for this class which explains what feature this tests, how to test it, and what the expected results are from said test.

Also, formatting: braces on newlines.
```suggestion
public class ServerWorldCreationTest
{
```

File: src/test/java/net/minecraftforge/debug/world/ServerWorldCreationTest.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    static final String MODID = "server_world_creation_test";
+    private static final Logger LOGGER = LogManager.getLogger(MODID);
```
Comment (sciwhiz12) [inline src/test/java/net/minecraftforge/debug/world/ServerWorldCreationTest.java]: As this logger is unused, please remove it and any unused imports.

File: src/test/java/net/minecraftforge/debug/world/ServerWorldCreationTest.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Minecraft Forge - Forge Development LLC
```
Comment (sciwhiz12) [inline src/test/java/net/minecraftforge/debug/world/ServerWorldCreationTest.java]: Outdated license header, please update to the latest commit of your base branch and run the `updateLicenses`/`licenseFormat` task.

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 6]
```diff
@@ -49,6 +52,8 @@
 import net.minecraft.ResourceLocationException;
 import net.minecraft.advancements.Advancement;
 import net.minecraft.core.*;
+import net.minecraft.resources.*;
```
Comment (sciwhiz12) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: Please avoid the use of wildcard imports.

File: src/main/resources/META-INF/accesstransformer.cfg [hunk 7]
```diff
@@ -263,6 +263,7 @@ protected net.minecraft.data.tags.TagsProvider m_6648_(Lnet/minecraft/resources/
 public net.minecraft.data.tags.TagsProvider$TagAppender f_126569_ # registry
 public net.minecraft.gametest.framework.GameTestServer <init>(Ljava/lang/Thread;Lnet/minecraft/world/level/storage/LevelStorageSource$LevelStorageAccess;Lnet/minecraft/server/packs/repository/PackRepository;Lnet/minecraft/server/WorldStem;Ljava/util/Collection;Lnet/minecraft/core/BlockPos;)V # constructor
 public net.minecraft.network.protocol.status.ClientboundStatusResponsePacket f_134885_ # GSON
+public net.minecraft.resources.RegistryLoader <init>(Lnet/minecraft/resources/RegistryResourceAccess;)V # <init>
```
Comment (sciwhiz12) [inline src/main/resources/META-INF/accesstransformer.cfg]: Please run the `checkATs` task to correct the comment of this entry and remove the trailing newline.

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 8]
```diff
@@ -1465,4 +1469,36 @@ public static MobEffect loadMobEffect(CompoundTag nbt, String key, @Nullable Mob
         }
     }
 
+    private static final ThreadLocal<RegistryResourceAccess> registryResouceAccess = new ThreadLocal<>();
```
Comment (sciwhiz12) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: Typo in field name; missing a `r` in `Resource`.
```suggestion
    private static final ThreadLocal<RegistryResourceAccess> registryResourceAccess = new ThreadLocal<>();
```

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 9]
```diff
@@ -1465,4 +1469,36 @@ public static MobEffect loadMobEffect(CompoundTag nbt, String key, @Nullable Mob
...
+    private static final ThreadLocal<RegistryAccess> registryAccess = new ThreadLocal<>();
+
+    public static RegistryAccess.Writable createAndCacheWritable(ResourceManager manager) {
```
Comment (sciwhiz12) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: Formatting: braces on newlines.
```suggestion
    public static RegistryAccess.Writable createAndCacheWritable(ResourceManager manager)
    {
```

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 10]
```diff
@@ -1465,4 +1469,36 @@ public static MobEffect loadMobEffect(CompoundTag nbt, String key, @Nullable Mob
...
+        DataResult<? extends Registry<DimensionType>> result = loader.overrideRegistryFromResources(registry, registry.key(), DimensionType.DIRECT_CODEC, RegistryOps.create(JsonOps.INSTANCE, writable));
+        result.error().ifPresent(err -> LOGGER.error("Unable to load Modded Dimension Types: {}", err.message()));
+
+        registryResouceAccess.set(resourceAccess);
+        registryAccess.set(writable);
+        return writable;
+    }
+
+    //Called from the tail of DimensionType#defaultDimensions
+    public static void registerDefaultDimensions(WritableRegistry<LevelStem> writableregistry) {
```
Comment (sciwhiz12) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: Formatting: braces on newlines.
```suggestion
    public static void registerDefaultDimensions(WritableRegistry<LevelStem> writableregistry)
    {
```

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 11]
```diff
@@ -1465,4 +1469,36 @@ public static MobEffect loadMobEffect(CompoundTag nbt, String key, @Nullable Mob
...
+        //This will always return unless #createAndCacheWritable has been called (we're on a dedicated server creating a new world)
+        if(registryAccess.get() == null || registryResouceAccess.get() == null) {
```
Comment (sciwhiz12) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: Formatting: space before an `if`-expression's opening parenthesis, and braces on newlines.
```suggestion
        if (registryAccess.get() == null || registryResouceAccess.get() == null)
        {
```

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]
I've not tested this PR yet (and my investigation through the code is stalled a bit since I have little knowledge of the world loading system). 

However, testing the code from the dropped patch (which I mentioned in the linked issue) shows that it works as expected. Is there a reason why that patch was not simply re-added by this PR?

If not, then please port over the patch instead as a simpler test. You may choose to keep or discard the test mod, but note that the issue can also be replicated with the `dimension_seed_test`'s dimensions.

<details>
<summary>Mapped and updated code of the dropped patch, for convenience</summary>

<p>

```java
// Forge: Deserialize the WorldGenSettings to ensure modded dims are loaded on first server load. Vanilla behavior only loads from the server.properties and deserializes only after the 2nd server load.
worldgensettings = WorldGenSettings.CODEC.encodeStart(dynamicops, worldgensettings).flatMap(nbt -> WorldGenSettings.CODEC.parse(dynamicops, nbt)).getOrThrow(false, errorMsg -> {});
```

</p>

</details>

---

Please remove the issue number from your title and amend your description to include the phrase `fixes #8552` or `closes #8552`, so the issue is closed when this PR is merged.

If you move to porting the patch over instead of your PR's additions, please mark the relevant review comments as resolved.

**Comment by Wyn-Price** [comment]
The original intent was to "inject" modded dimensions into the world gen settings. If the writing/parsing way works and is preferred, I can modify this PR to do that, or somebody can create a new PR. I admit, it is a much cleaner looking solution, albeit it marginally slower, as we have to parse all vanilla entries twice. 

The point of the test mod was to ensure that dim types, dims, and everything else (tested with noise settings) is also loaded correctly.

**Review by sciwhiz12 — APPROVED** [review]

**Pull request merged**